sql/core/src/main/scala/org/apache/spark/sql/execution/stat/FrequentItems.scala

```diff
@@ -38,13 +37,22 @@ private[sql] object FrequentItems extends Logging {
     def add(key: Any, count: Long): this.type = {
       if (baseMap.contains(key))  {
         baseMap(key) += count
+        if (baseMap(key) < minCount) minCount = baseMap(key)
       } else {
         if (baseMap.size < size) {
           baseMap += key -> count
+          if (count < minCount) minCount = count
         } else {
-          // TODO: Make this more efficient... A flatMap?
-          baseMap.retain((k, v) => v > count)
-          baseMap.transform((k, v) => v - count)
+          val remainder = count - minCount
+          if (remainder >= 0) {
+            baseMap += key -> count // something will get kicked out, so we can add this
+            baseMap.retain((k, v) => v > minCount)
+            baseMap.transform((k, v) => v - minCount)
+            minCount = baseMap.foldLeft(Long.MaxValue)((a, b) => math.min(a, b._2))
```

Comment:
Since we still need to recompute `minCount` after each update, I think the code is simpler without tracking `minCount` in all cases. We can compute `minCount` directly in this branch.

~~~scala
if (baseMap.contains(key)) {
  baseMap(key) += count
} else {
  baseMap(key) = count
  if (baseMap.size > size) {
    val minCount = baseMap.values.min
    baseMap.retain((_, v) => v > minCount)
    baseMap.transform((_, v) => v - minCount)
  }
}
~~~